meteor/server/api/playout/adlib.ts

```diff
@@ -411,6 +411,56 @@ export namespace ServerPlayoutAdLibAPI {
 		})
 	}
 
+	export function innerFindLastScriptedPieceOnLayer(
+		cache: CacheForPlayout,
+		sourceLayerId: string[],
+		customQuery?: MongoQuery<Piece>
+	) {
+		const span = profiler.startSpan('innerFindLastScriptedPieceOnLayer')
+
+		const playlist = cache.Playlist.doc
+		const rundownIds = getRundownIDsFromCache(cache)
+
+		if (!playlist.currentPartInstanceId) {
+			return
+		}
+
+		const currentPartInstance = cache.PartInstances.findOne(playlist.currentPartInstanceId)
+
+		if (!currentPartInstance) {
+			return
+		}
+
+		const query = {
+			...customQuery,
+			startRundownId: { $in: rundownIds },
+			sourceLayerId: { $in: sourceLayerId },
+		}
+
+		const pieces = Pieces.find(query).fetch()
```

Comment:
Looking at this again, I am a little concerned about the volume of data here. As you get further through the rundowns, the number of pieces this will find will keep on increasing.

The infinite logic does work similarly, but slightly more scoped (only certain infinite modes). Might be enough of a precedent to leave this be for now.
Or it could fetch only the id and startPartId fields, then do a second query to load the full copy of the pieces in the chosen part?

Reply:
Yeah I was concerned about that too, maybe it should be scoped to look back `n` parts? Where `n` can be specified by the adlib action or defaults to 10 or so?

Reply:
Maybe. That could be added later if this is found to be slow